src/runtime_src/core/pcie/driver/linux/xocl/xocl_subdev.c

```diff
@@ -462,6 +463,18 @@ static int __xocl_subdev_construct(xdev_handle_t xdev_hdl,
 					&res[i], subdev->info.level);
 			if (retval && retval != -ENODEV)
 				goto error;
+
+			/* If any addressable end point is of 64 bit xrt has
+			 * to pass only offset.
+			 * Get the offset by comparing with bar mappings of CPM.
+			 */
+			if(core->bars) {
+				bar_start = core->bars[bar_idx].base_addr;
+				if ((bar_start >> 32) == (res[i].start >> 32))
+					res[i].start ^= bar_start;
+				if ((bar_start >> 32) == (res[i].end >> 32))
+					res[i].end ^= bar_start;
+			}
```

Comment:
I am confused. 
(1) Is low 32-bit of bar_start always 0? Should it be res[i].start -= bar_start; 
(2) Why do we have to check if ((bar_start >> 32) == (res[i].start >> 32))? I think as long as BAR metadata shows up, we should do adjust the endpoint offset?

Reply:
As discussed offline, we will simplify this code for better understanding.

Reply:
@houlz0507 simplified the code and also added code to filter out PF0 info from userpf fdt blob, please review